### DISCUSSION TITLE
Py.test doesn't show local "caught exception" variables in detailed tracebacks

### DISCUSSION THREAD

**Discussion by pakal** [pytest-dev/pytest#2919] [General] [answered]
Hello,

I've noticed that when using showlocals ("-l" flag), pytest didnt display local exception variables created by an "except XXX as xxx" clause. 

It's however very useful to see the values of intermediate exceptions happening in frames of the traceback.

Pytest modules installed:
pytest==3.2.2
pytest-capturelog==0.7
pytest-django==3.1.2
pytest-vcr==0.3.0

Example of traceback output with "py.test -vl" ("e" is missing in showlocals part):

```
        except recurly.errors.NotFoundError as e: 
            if "Account" in e.message:
>               raise RuntimeError("Abnormal NotFoundError encountered in refresh of user %s" % user)
E               RuntimeError: Abnormal NotFoundError encountered in refresh of user testage_1510586801

account_code = 'testage_1510586801'
user       = <UserMedici: testage_1510586801>

b2c/recurly/core.py:115: RuntimeError
```

**Comment by nicoddemus** [comment]
Thanks @pakal for the report!

**Comment by symonk** [comment]
@nicoddemus any advice on where this is best fixed?

I have mocked a stub test for now:

```python
def test_exceptions_are_displayed_when_showlocals_is_passed(pytester: Pytester) -> None:
    pytester.makepyfile(
        """
        def foobar():
            try:
                x = 1
                y = 'foo'
                10 / 0
            except ZeroDivisionError as e:
                c = 'bar'
                raise RuntimeError(e)
                
        def test_it():
            foobar()
        """
    )
    result = pytester.runpytest("-v", "-l")
    assert result.ret == ExitCode.TESTS_FAILED
    result.stdout.fnmatch_lines([
        "x*= 1",
        "y*= 'foo'",
        "c*= 'bar'"
        #TODO: Ensure {'e'} is listed as in local scope.
    ])
```

`sys.exc_info()` looks kinda gobbled inside `ExceptionInfo.from_current`.  Any advice on where we are managing the `locals()` would be appreciated :)

**Comment by nicoddemus** [comment, marked as answer]
> @nicoddemus any advice on where this is best fixed?

Hmm unfortunately no, the exception variable is automatically deleted once the stack is unwound to avoid cyclic references... 

there's a workaround, which is to assign the exception to a different variable:

```python
        except recurly.errors.NotFoundError as e: 
            if "Account" in e.message:
                error = e
                raise RuntimeError("Abnormal NotFoundError encountered in refresh of user %s" % user)
```

But I don't know how it could be fixed.

**Comment by pakal** [comment]
So it's the Python runtime itself which deletes the exception?

I'm confused, isn't the GC able to deal with cyclic refererences except in rare cases?

**Comment by nicoddemus** [comment]
> So it's the Python runtime itself which deletes the exception?

Exactly. 👍 

> I'm confused, isn't the GC able to deal with cyclic refererences except in rare cases?

Yes but leaving that in there would create a cycle every time, and I guess that's something they want to avoid. Can't find more information about this right now, but I'm sure there are better explanations out there. 😁

**Comment by pakal** [comment]
Allright, I guess it's not something Pytest can fix, it would modify Python behaviour too much  ^^'

Is there someplace where it would be worth documenting this?
For example should I add a note in https://docs.pytest.org/en/stable/usage.html?highlight=show%20locals#modifying-python-traceback-printing or is it too specific?

**Comment by nicoddemus** [comment]
A note there might be worthwhile yes! 👍
